contracts/access/manager/AccessManager.sol

```diff
@@ -0,0 +1,564 @@
+     * @dev Internal version of {setFunctionAllowedGroup} without access control.
+     *
+     * Emit a {FunctionAllowedGroupUpdated} event
+     */
+    function _setFunctionAllowedGroup(address target, bytes4 selector, uint256 groupId) internal virtual {
+        _allowedGroups[target][selector] = groupId;
+        emit FunctionAllowedGroupUpdated(target, selector, groupId);
+    }
+
+    // =============================================== MODE MANAGEMENT ================================================
+    /**
+     * @dev Set the operating mode of a contract to Custom. This enables the group mechanism for per-function access
+     * restriction and delay enforcement.
+     *
+     * Requirements:
+     *
+     * - the caller must be a global admin
+     *
+     * Emits a {AccessModeUpdated} event.
+     */
+    function setContractModeCustom(address target) public virtual onlyGroup(ADMIN_GROUP) {
+        // todo set delay or document risks
+        _setContractMode(target, AccessMode.Custom);
+    }
+
+    /**
+     * @dev Set the operating mode of a contract to Open. This allows anyone to call any `restricted()` function with
+     * no delay.
+     *
+     * Requirements:
+     *
+     * - the caller must be a global admin
+     *
+     * Emits a {AccessModeUpdated} event.
+     */
+    function setContractModeOpen(address target) public virtual onlyGroup(ADMIN_GROUP) {
+        // todo set delay or document risks
+        _setContractMode(target, AccessMode.Open);
+    }
+
+    /**
+     * @dev Set the operating mode of a contract to Close. This prevents anyone from calling any `restricted()`
+     * function.
+     *
+     * Requirements:
+     *
+     * - the caller must be a global admin
+     *
+     * Emits a {AccessModeUpdated} event.
+     */
+    function setContractModeClosed(address target) public virtual onlyGroup(ADMIN_GROUP) {
+        // todo set delay or document risks
+        _setContractMode(target, AccessMode.Closed);
+    }
+
+    /**
+     * @dev Set the operating mode of a contract. This is an internal setter with no access restrictions.
+     *
+     * Emits a {AccessModeUpdated} event.
+     */
+    function _setContractMode(address target, AccessMode mode) internal virtual {
+        _contractMode[target] = mode;
+        emit AccessModeUpdated(target, mode);
+    }
+
+    // ============================================== DELAYED OPERATIONS ==============================================
+    /**
+     * @dev Return the timepoint at which an operation was scheduled.
+     */
+    function getSchedule(bytes32 id) public view virtual returns (uint48) {
+        return _schedules[id];
+    }
+
+    /**
+     * @dev Schedule a delayed operation, and return the operation identifier.
+     *
+     * Emits a {Scheduled} event.
+     */
+    function schedule(address target, bytes calldata data) public virtual returns (bytes32) {
+        address caller = _msgSender();
+        bytes4 selector = bytes4(data[0:4]);
```

Comment:
All of these `data[0:4]` should be:

```suggestion
        bytes4 selector = bytes4(msgData()[0:4]);
```

Otherwise we'd have a mismatch with the sender if this contract is using the ERC2771Forwarder

Reply:
Ok I realized this is wrong completely.